### PR TITLE
ASN.1: Add sanity checks for input len <= 0 and input it == NULL; related coding improvements

### PR DESCRIPTION
**Pull request by DDvO**
This has been motivated by a crash on error reported in #14322.

### CODE REVIEW

File: crypto/asn1/a_d2i_fp.c [hunk 1]
```diff
@@ -138,15 +140,18 @@ int asn1_d2i_read_bio(BIO *in, BUF_MEM **pb)
...
+        if (diff == 0)
+            goto err;
+        inf = ASN1_get_object(&q, &slen, &tag, &xclass, diff);
         if (inf & 0x80) {
             unsigned long e;
 
             e = ERR_GET_REASON(ERR_peek_error());
             if (e != ASN1_R_TOO_LONG)
                 goto err;
             else
```
Comment (richsalz) [inline crypto/asn1/a_d2i_fp.c:153]: else after goto is bad style because you're just falling through to the next lines anyway

Comment (DDvO) [reply crypto/asn1/a_d2i_fp.c]: Right - so I've cleaned this up as well.

(There would be also deeper things to clean up, such as the use of needlessly non-conservative `ERR_clear_error()`, but that's out of scope of this PR.)

File: crypto/asn1/tasn_dec.c [hunk 2]
```diff
@@ -1119,25 +1135,22 @@ static int asn1_check_tlen(long *olen, int *otag, unsigned char *oclass,
...
         }
     }
 
-    if (i & 0x80) {
-        ERR_raise(ERR_LIB_ASN1, ASN1_R_BAD_OBJECT_HEADER);
-        asn1_tlc_clear(ctx);
-        return 0;
-    }
+    if ((i & 0x80) != 0)
+        goto err;
```
Comment (t8m) [inline crypto/asn1/tasn_dec.c:1147]: Should there be just asn1_tlc_clear() in the err: block so you could reuse it by the other various return 0 calls here.

File: crypto/asn1/tasn_dec.c [hunk 3]
```diff
@@ -1100,15 +1112,19 @@ static int asn1_check_tlen(long *olen, int *otag, unsigned char *oclass,
     p = *in;
     q = p;
 
-    if (ctx && ctx->valid) {
+    if (len <= 0) {
+        ERR_raise(ERR_LIB_ASN1, ASN1_R_TOO_SMALL);
```
Comment (t8m) [inline crypto/asn1/tasn_dec.c:1116]: You have double ERR_raise() calls here.

File: crypto/asn1/tasn_dec.c (hunk 2 above)
Comment (DDvO) [reply crypto/asn1/tasn_dec.c:1147]: Good thought - done.
This also fixes the issue below.

File: crypto/asn1/tasn_dec.c (hunk 3 above)
Comment (DDvO) [reply crypto/asn1/tasn_dec.c:1116]: Fixed by the above suggestion.

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Merged - thanks @t8m

**Pull request closed without merging**